src/hotspot/share/prims/universalUpcallHandler.cpp

```diff
@@ -60,20 +60,100 @@ Thread* ProgrammableUpcallHandler::maybe_attach_and_get_thread(bool* should_deta
     guarantee(result == JNI_OK, "Could not attach thread for upcall. JNI error code: %d", result);
     *should_detach = true;
     thread = Thread::current();
```

Comment:
You could use JavaThread::current() here and avoid the later conversions.

Reply:
Ok. The variable is initially initialized (hah) using `Thread::current_or_null()` and there was no equivalent that does that for `JavaThread`. I can add something like that though.